src/core/shared_memory_manager.cc

```diff
@@ -0,0 +1,244 @@
+    void** shm_addr)
+{
+  // map shared memory to process address space
+  *shm_addr = mmap(NULL, byte_size, PROT_WRITE, MAP_SHARED, shm_fd, offset);
+  if (*shm_addr == MAP_FAILED) {
+    return Status(
+        RequestStatusCode::INTERNAL, "Unable to process address space");
+  }
+
+  return Status::Success;
+}
+
+Status
+CloseSharedMemoryRegion(int shm_fd)
+{
+  int tmp = close(shm_fd);
+  if (tmp == -1) {
+    return Status(
+        RequestStatusCode::INTERNAL, "Unable to close shared memory region");
+  }
+
+  return Status::Success;
+}
+
+Status
+MmapCleanup(void* shm_addr, size_t byte_size)
+{
+  int tmp_fd = munmap(shm_addr, byte_size);
+  if (tmp_fd == -1) {
+    return Status(
+        RequestStatusCode::INTERNAL, "Unable to munmap shared memory region");
+  }
+
+  return Status::Success;
+}
+
+}  // namespace
+
+Status
+SharedMemoryManager::RegisterSharedMemory(
+    const std::string& name, const std::string& shm_key, const size_t offset,
+    const size_t byte_size)
+{
+  LOG_VERBOSE(1) << "Register() shared memory region: '" << name << "'";
+
+  // Serialize all operations that write/read current shared memory regions
+  std::lock_guard<std::mutex> lock(register_mu_);
+
+  // If key is already in shared_memory_map_ then return error saying already
+  // registered
+  if (shared_memory_map_.find(name) != shared_memory_map_.end()) {
+    return Status(
+        RequestStatusCode::INVALID_ARG,
+        "The shared memory region '" + name + "' is already registered.");
+  }
+
+  // register (or re-register)
+  try {
+    void* tmp_addr;
+    int shm_fd = -1;
+
+    // don't re-open if shared memory is already open
+    for (auto itr = shared_memory_map_.begin(); itr != shared_memory_map_.end();
+         ++itr) {
+      if (itr->second->shm_key_ == shm_key)
+        shm_fd = itr->second->shm_fd_;
+      else {
+        RETURN_IF_ERROR(OpenSharedMemoryRegion(shm_key, &shm_fd));
+      }
+    }
+
+    RETURN_IF_ERROR(
+        GetSharedMemoryAddress(shm_fd, offset, byte_size, &tmp_addr));
+    shared_memory_map_.insert(std::make_pair(
+        name, std::unique_ptr<SharedMemoryInfo>(new SharedMemoryInfo(
+                  name, shm_key, offset, byte_size, shm_fd, tmp_addr))));
+  }
+  catch (std::exception& ex) {
+    return Status(
+        RequestStatusCode::INVALID_ARG,
+        "Failed to register shared memory region '" + name + "'.");
+  }
+
+  return Status::Success;
+}
+
+Status
+SharedMemoryManager::UnregisterSharedMemory(const std::string& name)
+{
+  LOG_VERBOSE(1) << "Unregister() shared memory region: '" << name << "'";
+
+  // Serialize all operations that write/read current shared memory regions
+  std::lock_guard<std::mutex> lock(register_mu_);
+
+  auto it = shared_memory_map_.find(name);
+  if (it != shared_memory_map_.end()) {
+    RETURN_IF_ERROR(MmapCleanup(it->second->shm_addr_, it->second->byte_size_));
+
+    // if no other region with same shm_key then close
```

Comment:
You need to remove the region from the map before this or you will always find it